File: backend/src/providers/commentProvider.ts
import { prisma } from '../lib/prisma.js';

/**
 * Data Access Layer — Comments & Replies
 */
export const commentProvider = {
  /** Return all visible, non-deleted comments for a menu item, newest first. */
  async findByMenuItemId(menuItemId: bigint) {
    return prisma.menu_item_comments.findMany({
      where: { menu_item_id: menuItemId, deleted: false, status: 'Visible' },
      include: {
        users: { select: { id: true, name: true, img: true } },
        menu_item_comment_replies: {
          where: { deleted: false },
          include: { users: { select: { id: true, name: true, img: true } } },
          take: 1,
        },
      },
      orderBy: { created: 'desc' },
    });
  },

  /** Find a single comment by id. */
  async findById(id: bigint) {
    return prisma.menu_item_comments.findFirst({
      where: { id, deleted: false },
    });
  },

  /** Create a new comment. */
  async create(data: {
    menuItemId: bigint;
    customerId: bigint;
    content: string;
    rating?: number;
  }) {
    return prisma.menu_item_comments.create({
      data: {
        menu_item_id: data.menuItemId,
        customer_id: data.customerId,
        content: data.content,
        rating: data.rating,
      },
    });
  },

  /** Find the first (non-deleted) reply for a comment. */
  async findReplyByCommentId(commentId: bigint) {
    return prisma.menu_item_comment_replies.findFirst({
      where: { comment_id: commentId, deleted: false },
    });
  },

  /** Create a reply to a comment. */
  async createReply(data: {
    commentId: bigint;
    staffId: bigint;
    content: string;
  }) {
    return prisma.menu_item_comment_replies.create({
      data: {
        comment_id: data.commentId,
        staff_id: data.staffId,
        content: data.content,
      },
      include: {
        users: { select: { id: true, name: true, img: true } },
      },
    });
  },

  /**
   * Return all comments (for staff/admin), newest first.
   * Includes the menu item name for display purposes.
   */
  async findAll(options: {
    limit?: number;
    status?: string;
    menuItemId?: bigint;
  } = {}) {
    return prisma.menu_item_comments.findMany({
      where: {
        deleted: false,
        ...(options.status ? { status: options.status } : {}),
        ...(options.menuItemId ? { menu_item_id: options.menuItemId } : {}),
      },
      include: {
        users: { select: { id: true, name: true, img: true } },
        menu_items: { select: { id: true, name: true } },
        menu_item_comment_replies: {
          where: { deleted: false },
          include: { users: { select: { id: true, name: true, img: true } } },
          take: 1,
        },
      },
      orderBy: { created: 'desc' },
      take: options.limit ?? 50,
    });
  },

  /** Toggle comment visibility status (Visible / Hidden). */
  async updateStatus(id: bigint, status: 'Visible' | 'Hidden') {
    return prisma.menu_item_comments.update({
      where: { id },
      data: { status },
    });
  },
};

File: backend/app/providers/comment_provider.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update
from sqlalchemy.orm import selectinload
from typing import Optional, List
from app.models.menu_item_comment import MenuItemComment
from app.models.menu_item_comment_reply import MenuItemCommentReply
from app.models.user import User

class CommentProvider:
    async def find_by_menu_item(self, db: AsyncSession, menu_item_id: int) -> List[MenuItemComment]:
        result = await db.execute(
            select(MenuItemComment)
            .options(selectinload(MenuItemComment.user), selectinload(MenuItemComment.replies).selectinload(MenuItemCommentReply.user))
            .where(MenuItemComment.menu_item_id == menu_item_id, MenuItemComment.deleted == False, MenuItemComment.status == "Visible")
            .order_by(MenuItemComment.created.desc())
        )
        return list(result.scalars().all())

    async def find_all(self, db: AsyncSession, status: Optional[str] = None, menu_item_id: Optional[int] = None, limit: int = 50) -> List[MenuItemComment]:
        stmt = select(MenuItemComment).options(selectinload(MenuItemComment.user), selectinload(MenuItemComment.replies)).where(MenuItemComment.deleted == False)
        if status:
            stmt = stmt.where(MenuItemComment.status == status)
        if menu_item_id:
            stmt = stmt.where(MenuItemComment.menu_item_id == menu_item_id)
        stmt = stmt.order_by(MenuItemComment.created.desc()).limit(limit)
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def create(self, db: AsyncSession, menu_item_id: int, customer_id: int, content: str, rating: Optional[float] = None) -> MenuItemComment:
        comment = MenuItemComment(menu_item_id=menu_item_id, customer_id=customer_id, content=content, rating=rating)
        db.add(comment)
        await db.commit()
        await db.refresh(comment)
        return comment

    async def create_reply(self, db: AsyncSession, comment_id: int, staff_id: int, content: str) -> MenuItemCommentReply:
        reply = MenuItemCommentReply(comment_id=comment_id, staff_id=staff_id, content=content)
        db.add(reply)
        await db.commit()
        await db.refresh(reply)
        return reply

    async def update_status(self, db: AsyncSession, comment_id: int, status: str) -> Optional[MenuItemComment]:
        await db.execute(update(MenuItemComment).where(MenuItemComment.id == comment_id).values(status=status))
        await db.commit()
        result = await db.execute(select(MenuItemComment).where(MenuItemComment.id == comment_id))
        return result.scalar_one_or_none()

    async def find_by_id(self, db: AsyncSession, comment_id: int) -> Optional[MenuItemComment]:
        result = await db.execute(select(MenuItemComment).where(MenuItemComment.id == comment_id))
        return result.scalar_one_or_none()

comment_provider = CommentProvider()
